File: LCMV-test/create_signals.m
clear all
close all
load handel.mat
addpath('C:\project\fillline.m');
%% variens

clear all
%this is the radios of the table.

c_k = 340;                                        % Sound velocity (m/s)
c = 340; 
fs = 16000;
n = 4096;                                         % Number of samples
mtype = 'omnidirectional';                        % Type of microphone
order = -1;                                       % -1 equals maximum reflection order!
dim = 3;                                          % Room dimension
orientation = 0;                                  % Microphone orientation (rad)
hp_filter = 1;                                    % Disable high-pass filter
% define analysis parameters
NO_S_i=10;
lottery=2000;
R=1.3;
R_small = 1.2;
pad=25000;
start=1;
max1=100;
noise_R=0.2;
flag=0;
flag2=0;
nfft=2048; 
hop=1024;
num_jumps=9;
high=1;
angle=360;
distance_from_woll=0.5;
radius_mics = 0.1;
%% rir generation & covariens sounds 

close all
% define room dimension
L1_temp=randi([1,20]);
L1=4+0.1*L1_temp;
L2_temp=randi([1,20]);
L2=4+0.1*L2_temp;
L = [L1 L2 3];
room_x = L(1);
room_y = L(2);
SNR_difuse=20;
beta=0.3;
SIR = 0;

% calc snr
SNR_direction = 20;
SNR_mic = 30;
%% create circle & line & mic location % speaker location

distance_total=R+distance_from_woll+noise_R;
end_point_x=room_x-(R+distance_from_woll+noise_R);
end_point_y=room_y-(R+distance_from_woll+noise_R);
Radius_X = (end_point_x-distance_total).*rand + distance_total;
Radius_Y = (end_point_y-distance_total).*rand + distance_total;

R_angle=randi([1,angle]); % take  rand 180 degrees of circle to create rand orientation of Microphone array
t = linspace(-pi,2*pi,angle+angle/2); % 3 times 180 to create option to rand whole circ
t=t(R_angle:R_angle+angle/2-1);
x = R*sin(t)+Radius_X;
y = R*cos(t)+Radius_Y;
x_small = R_small*sin(t)+Radius_X;
y_small = R_small*cos(t)+Radius_Y;

z=0*t+high;
circ_mics_x = radius_mics*sin(t)+Radius_X;
circ_mics_y = radius_mics*cos(t)+Radius_Y;
[line_x,line_y]=fillline([x(1) y(1)], [x(angle/2) y(angle/2)],R*2*100);

r= [circ_mics_x(1) circ_mics_y(1) high; circ_mics_x(60) circ_mics_y(60) high;...
    circ_mics_x(120) circ_mics_y(120) high; circ_mics_x(180) circ_mics_y(180) high];

%% create locations   
    
center=[Radius_X Radius_Y];
start_circ_vec=[line_x(1) line_y(1)]-center;
labels_location = 5:10:175;
list_locations = [];
next_speech = 1;
while next_speech
    next_speech = 0;
    rand1 = randi(angle/2); 
    x1 = x(rand1);
    y1 = y(rand1);
    x1_temp=x1;
    y1_temp=y1;
    w=0.01*randi([1,314]);
    x1=x1+noise_R*sin(w);
    y1=y1+noise_R*cos(w);

    rand2 = randi(angle/2); 
    x2 = x(rand2);
    y2 = y(rand2);
    x2_temp=x2;
    y2_temp=y2;
    w=0.01*randi([1,314]);
    x2=x2+noise_R*sin(w);
    y2=y2+noise_R*cos(w);

    loc_xy = [x1,y1;x2,y2];
    dist = pdist(loc_xy,'euclidean');
    if dist<0.5
        next_speech = 1;
    end
end

s_first = [x1 ; y1 ; high];
s_second = [x2 ; y2 ; high];

%% create speakers clean

male_female_check=randi([1,2]);

if male_female_check==1
   male_female1='male'; 
   NO_S=112;
else
   male_female1='female'; 
   NO_S=56;
end

source=randi([1,NO_S]);
source1 = int2str(source);

male_female_check=randi([1,2]);

if male_female_check==1
   male_female2='male'; 
   NO_S=112;
else
   male_female2='female'; 
   NO_S=56;
end

source=randi([1,NO_S]);
source2 = int2str(source);    


path1 = strcat('C:\project\TIMIT CD\TIMIT\test\',male_female1,'\',male_female1,'_',source1);
path2 = strcat('C:\project\TIMIT CD\TIMIT\test\',male_female2,'\',male_female2,'_',source2);

f1=dir([path1 '\*.wav']);
f2=dir([path2 '\*.wav']);
folder1=f1.folder;
folder2=f2.folder;

file11=f1(3).name;
file12=f1(4).name;

file21=f1(5).name;
file22=f1(6).name;

source11 = strcat(folder1,'\',file11);
source12 = strcat(folder1,'\',file12);
source21 = strcat(folder1,'\',file21);
source22 = strcat(folder1,'\',file22);

speech_11_temp = audioread(source11);
speech_12_temp = audioread(source12);
speech_21_temp = audioread(source21);
speech_22_temp = audioread(source22);

speech_11=speech_11_temp(:,1);
speech_12=speech_12_temp(:,1);
speech_21=speech_21_temp(:,1);
speech_22=speech_22_temp(:,1);


pad_zeros1=zeros(1,fs)';
pad_zeros2=zeros(1,fs*2)';

in1 = [pad_zeros2 ; speech_11 ; pad_zeros1 ; zeros(1,length(speech_21))' ; pad_zeros1 ; speech_12];
vad1 = [pad_zeros2 ; ones(1,length(speech_11))' ; pad_zeros1 ; zeros(1,length(speech_21))' ; pad_zeros1 ; ones(1,length(speech_12))'];

in2 = [pad_zeros2 ; zeros(1,length(speech_11))' ; pad_zeros1 ; speech_21 ; pad_zeros1 ; speech_22];
vad2 = [pad_zeros2 ; zeros(1,length(speech_11))' ; pad_zeros1 ; ones(1,length(speech_21))' ; pad_zeros1 ; ones(1,length(speech_22))'];

maxlen = max(length(in1), length(in2));
in1(end+1:maxlen,:) = 0;
in2(end+1:maxlen,:) = 0;
vad1(end+1:maxlen,:) = 0;
vad2(end+1:maxlen,:) = 0;

in1 = in1.';
in2 = in2.';

vad1 = vad1.';
vad2 = vad2.';

vad = vad1+vad2;
 

%% create RIRs

h_first = rir_generator(c_k, fs, r, s_first', L, beta, n, mtype, order, dim, orientation, hp_filter);    
receiver_first = conv2(in1,h_first);

h_second = rir_generator(c_k, fs, r, s_second', L, beta, n, mtype, order, dim, orientation, hp_filter);    
receiver_second = conv2(in2,h_second);

%% create noise

middle = [Radius_X Radius_Y high];
s_noise = [Radius_X Radius_Y high];
d_noise = norm(s_noise-middle);
while d_noise<2
    x_noise=distance_from_woll+0.01*randi(100)*(room_x-2*distance_from_woll);
    y_noise=distance_from_woll+0.01*randi(100)*(room_y-2*distance_from_woll);
    s_noise = [x_noise y_noise high];
    d_noise = norm(s_noise-middle);
end

p=randi([1,5]);
noise_file=sprintf('C:\\project\\noise\\noise_%d.wav',p);
noise_temp = audioread(noise_file);
noise_temp = [noise_temp.' noise_temp.' noise_temp.'].';
noise_temp=noise_temp(1:length(receiver_first)-n+1)';
h_noise = rir_generator(c_k, fs, r, s_noise, L, beta, n, mtype, order, dim, orientation, hp_filter);
Receivers_noise = conv2(noise_temp,h_noise);    
Receivers_noise = Receivers_noise/mean(std(Receivers_noise'));
%% create input

M=size(receiver_first,1);
length_receives=size(receiver_first,2);
difuse_noise = create_difuse_noise(length_receives);
difuse_noise = difuse_noise/mean(std(difuse_noise));
receiver_first=receiver_first.'/max(max(abs(receiver_first.')));
receiver_second=receiver_second.'/max(max(abs(receiver_second.'))); 

A_x_before=(mean(std(receiver_first))+mean(std(receiver_second)))/2;
Ax_SIR = A_x_before/(10^(SIR/20));
receivers = Ax_SIR*receiver_first.'+receiver_second.';
%clac An
A_x=mean(std(receivers'));
std_n=mean(std(Receivers_noise'));
A_n_difuse = A_x/(10^(SNR_difuse/20));
A_n_diraction = A_x/(10^(SNR_direction/20));
A_n_mic = A_x/(10^(SNR_mic/20));

%create mic moise
mic_noise = A_n_mic*randn(M,length_receives);    
% create difuse noise
receivers=(receivers+mic_noise+A_n_difuse*difuse_noise.'+A_n_diraction*Receivers_noise).'; 
noise_only = (mic_noise+A_n_difuse*difuse_noise.'+A_n_diraction*Receivers_noise).';
receivers=receivers/max(max(abs(receivers)));     

path = sprintf('C:\\Users\\user\\Desktop\\master\\Thesis\\sharon\\pini\\');
first_clean_file=strcat(path,'first_clean.wav');
second_clean_file=strcat(path,'second_clean.wav');
first_reverb_file=strcat(path,'first_reverb.wav');
second_reverb_file=strcat(path,'second_reverb.wav');
data_noise=strcat(path,'noise.wav');

audiowrite(first_clean_file,in1,fs) 
audiowrite(second_clean_file,in2,fs) 
audiowrite(first_reverb_file,receiver_first,fs) 
audiowrite(second_reverb_file,receiver_second,fs) 
audiowrite(data_noise,noise_only,fs) 

File: LCMV-Python/create_signals.py
from create_signals_functions import *
import matplotlib.pyplot as plt
import rir_generator as rir
import soundfile as sf
import scipy.signal as ss
import numpy as np
import os

# Ensure 'wav_files' directory exists in the same directory as this script
# Place  in the 'wav_files' directory the required audio files:
#   (male_11.wav, male_12.wav, female_21.wav, female_22.wav)


# ==========================================
# 1. HYPERPARAMETER DICTIONARY
# ==========================================
hParams = {
    # Physics & Environment
    'seed': 42,                    # Random seed for reproducibility
    'sound_speed': 340.0,
    'fs': 44100,                  # Target sampling rate
    'room_height': 3,
    'room_dim_min': 4.1,          # Minimum length/width for random room
    'room_dim_max': 6.0,          # Maximum length/width
    'rt60': 0.3,                  # Reverberation time (beta)
    
    # RIR Generation
    'rir_length': 4096,           # Samples (n)
    'mic_type': rir.mtype.omnidirectional,
    'order': -1,                  # Reflection order (-1 is max)
    'dim': 3,                     # 3D simulation
    'orientation': [0, 0],        # [Azimuth, Elevation]
    'hp_filter': True,            # Enable High-pass filter
    
    # Geometry (Mics & Speakers)
    'array_center_dist': 1.3,     # Distance of table from random center (R)
    'array_radius': 0.1,          # Radius of the mic array (radius_mics)
    'mic_height': 1.0,            # Height of mics (high)
    'wall_margin': 0.5,           # Min distance from wall (distance_from_woll)
    'speaker_min_dist': 0.5,      # Min distance between speakers
    'noise_source_dist': 2.0,     # Min distance for noise source from center
    'noise_var': 0.2,             # Variance for random placement (noise_R)
    
    # Microphone Indices (Selecting 4 mics from the 360 generated points)
    'mic_indices': [0, 59, 119, 179], 
    
    # Mixing Ratios (dB)
    'sir': -20,                     # Signal-to-Interference Ratio
    'snr_diffuse': 20,            # Diffuse (Ambient) Noise
    'snr_directional': 2,        # Directional Noise
    'snr_mic': 30                 # Microphone (Sensor) Noise
}

# Seeded RNG
rng = np.random.default_rng(hParams['seed'])

# ==========================================
# 2. MAIN SCRIPT
# ==========================================

# Setup Paths
current_dir = os.path.dirname(os.path.abspath(__file__))
wav_dir = os.path.join(current_dir, "wav_files")

# Verify Sampling Rate matches hParams
# (Optional: You could also resample here if they don't match)
_, file_fs = sf.read(os.path.join(wav_dir, "female_21.wav"))
if file_fs != hParams['fs']:
    print(f"Warning: File sampling rate ({file_fs}) differs from hParams ({hParams['fs']})")

# --- Geometry Setup ---

# Randomize Room Dimensions
L1 = 4 + 0.1 * rng.integers(
    (hParams['room_dim_min']-4)*10,
    (hParams['room_dim_max']-4)*10 + 1
)
L2 = 4 + 0.1 * rng.integers(
    (hParams['room_dim_min']-4)*10,
    (hParams['room_dim_max']-4)*10 + 1
)
L = [L1, L2, hParams['room_height']]
room_x, room_y = L[0], L[1]

# Calculate Safe Zone for Array Placement
# We define the "end points" to ensure the array fits within the room with margins
dist_total = hParams['array_center_dist'] + hParams['wall_margin'] + hParams['noise_var']
end_point_x = room_x - dist_total
end_point_y = room_y - dist_total

# Randomize center of the array within safe zone
Radius_X = (end_point_x - dist_total) * rng.random() + dist_total
Radius_Y = (end_point_y - dist_total) * rng.random() + dist_total

# Randomize Array Orientation
angle = 360
R_angle = rng.integers(0, angle) 
t_full = np.linspace(-np.pi, 2 * np.pi, int(angle + angle/2)) 
t = t_full[R_angle : R_angle + int(angle/2)]

# Calculate Coordinates
# Table center offset (x, y)
x = hParams['array_center_dist'] * np.sin(t) + Radius_X
y = hParams['array_center_dist'] * np.cos(t) + Radius_Y

# Mic locations (circ_mics)
circ_mics_x = hParams['array_radius'] * np.sin(t) + Radius_X
circ_mics_y = hParams['array_radius'] * np.cos(t) + Radius_Y

# Line visualization (Helper)
line_x, line_y = fillline(
    [x[0], y[0]], 
    [x[int(angle/2)-1], y[int(angle/2)-1]], 
    hParams['array_center_dist']*2*100
)

# Select specific microphones from the circle
indices = [i for i in hParams['mic_indices'] if i < len(circ_mics_x)]
r = np.zeros((len(indices), 3))
for i, idx in enumerate(indices):
    r[i, :] = [circ_mics_x[idx], circ_mics_y[idx], hParams['mic_height']]

# Place Speakers (Source Locations)
next_speech = True
while next_speech:
    next_speech = False
    
    # Randomize Speaker 1
    rand1 = rng.integers(0, int(angle/2))
    w1 = 0.01 * rng.integers(1, 315)
    x1 = x[rand1] + hParams['noise_var'] * np.sin(w1)
    y1 = y[rand1] + hParams['noise_var'] * np.cos(w1)

    # Randomize Speaker 2
    rand2 = rng.integers(0, int(angle/2))
    w2 = 0.01 * rng.integers(1, 315)
    x2 = x[rand2] + hParams['noise_var'] * np.sin(w2)
    y2 = y[rand2] + hParams['noise_var'] * np.cos(w2)

    # Check Minimum Distance
    loc_xy = np.array([[x1, y1], [x2, y2]])
    dist = np.linalg.norm(loc_xy[0] - loc_xy[1])
    
    if dist < hParams['speaker_min_dist']:
        next_speech = True

s_first = [x1, y1, hParams['mic_height']]
s_second = [x2, y2, hParams['mic_height']]

# --- Prepare Audio Signals ---

speech_11 = load_clean_wav(os.path.join(wav_dir, "male_11.wav"))
speech_12 = load_clean_wav(os.path.join(wav_dir, "male_12.wav"))
speech_21 = load_clean_wav(os.path.join(wav_dir, "female_21.wav"))
speech_22 = load_clean_wav(os.path.join(wav_dir, "female_22.wav"))

fs = hParams['fs']
pad_zeros1 = np.zeros(fs) 
pad_zeros2 = np.zeros(fs * 2) 

# Timeline Construction
# Timeline 1: [Silence (2s)] [Spk1] [Silence] [Silence] [Silence] [Spk1]
in1 = np.concatenate([
    pad_zeros2, speech_11, pad_zeros1, np.zeros(len(speech_21)), pad_zeros1, speech_12
])
# Timeline 2: [Silence (2s)] [Silence] [Silence] [Spk2] [Silence] [Spk2]
in2 = np.concatenate([
    pad_zeros2, np.zeros(len(speech_11)), pad_zeros1, speech_21, pad_zeros1, speech_22
])

# Pad to match lengths
maxlen = max(len(in1), len(in2))
in1 = np.pad(in1, (0, maxlen - len(in1)), 'constant')
in2 = np.pad(in2, (0, maxlen - len(in2)), 'constant')

# --- Room Simulation (RIR) ---

common_rir_args = {
    'c': hParams['sound_speed'], 
    'fs': hParams['fs'], 
    'r': r, 
    'L': L, 
    'reverberation_time': hParams['rt60'], 
    'nsample': hParams['rir_length'], 
    'mtype': hParams['mic_type'], 
    'order': hParams['order'], 
    'dim': hParams['dim'], 
    'orientation': hParams['orientation'], 
    'hp_filter': hParams['hp_filter']
}

# Generate RIRs
h_first = rir.generate(s=s_first, **common_rir_args)
h_second = rir.generate(s=s_second, **common_rir_args)

# Convolve
receiver_first = ss.fftconvolve(in1[:, np.newaxis], h_first, mode='full')
receiver_second = ss.fftconvolve(in2[:, np.newaxis], h_second, mode='full')

# --- Noise Generation ---

# 1. Directional Noise Placement
middle = np.array([Radius_X, Radius_Y, hParams['mic_height']])
s_noise = np.array([Radius_X, Radius_Y, hParams['mic_height']])
d_noise = 0

while d_noise < hParams['noise_source_dist']:
    x_noise = hParams['wall_margin'] + 0.01 * rng.integers(0, 100) * (room_x - 2 * hParams['wall_margin'])
    y_noise = hParams['wall_margin'] + 0.01 * rng.integers(0, 100) * (room_y - 2 * hParams['wall_margin'])
    s_noise = np.array([x_noise, y_noise, hParams['mic_height']])
    d_noise = np.linalg.norm(s_noise - middle)

# 2. Generate Directional Noise Signal
noise_len_needed = receiver_first.shape[0] - hParams['rir_length'] + 1
noise_temp_mono = rng.standard_normal(noise_len_needed)
noise_temp = np.tile(noise_temp_mono[:, np.newaxis], (1, 1))

# 3. Process Directional Noise
h_noise = rir.generate(s=s_noise, **common_rir_args)
Receivers_noise = ss.fftconvolve(noise_temp, h_noise, mode='full')

# Normalize Directional Noise
std_vals = np.std(Receivers_noise, axis=1)
if np.mean(std_vals) != 0:
    Receivers_noise = Receivers_noise / np.mean(std_vals)

# 4. Generate Diffuse Noise
M = receiver_first.shape[1] 
length_receives = receiver_first.shape[0]

difuse_noise = create_difuse_noise(length_receives, M=M, Fs=fs, c=hParams['sound_speed'])

if np.mean(np.std(difuse_noise)) != 0:
    difuse_noise = difuse_noise / np.mean(np.std(difuse_noise))

# --- Mixing ---

# Normalize Reverb Signals
receiver_first = receiver_first / np.max(np.abs(receiver_first))
receiver_second = receiver_second / np.max(np.abs(receiver_second))

# Apply SIR (Speaker Balance)
A_x_before = (np.mean(np.std(receiver_first)) + np.mean(np.std(receiver_second))) / 2
Ax_SIR = A_x_before / (10**(hParams['sir']/20.0))
receivers_speech = Ax_SIR * receiver_first + receiver_second

# Determine overall level for Noise Scaling
A_x = np.mean(np.std(receivers_speech))

# Align Lengths
min_len = min(receivers_speech.shape[0], Receivers_noise.shape[0], difuse_noise.shape[0])
receivers_speech = receivers_speech[:min_len, :]
Receivers_noise = Receivers_noise[:min_len, :]
difuse_noise = difuse_noise[:min_len, :]

# Calculate Noise Amplitudes
A_n_difuse = A_x / (10**(hParams['snr_diffuse']/20.0))
A_n_direction = A_x / (10**(hParams['snr_directional']/20.0))
A_n_mic = A_x / (10**(hParams['snr_mic']/20.0))

# 5. Generate Microphone Noise
mic_noise = A_n_mic * rng.standard_normal((min_len, M))

# Final Mix
receivers_mixed = (receivers_speech + 
                   mic_noise + 
                   A_n_difuse * difuse_noise + 
                   A_n_direction * Receivers_noise)

noise_only = (mic_noise + 
              A_n_difuse * difuse_noise + 
              A_n_direction * Receivers_noise)

# Final Normalization
receivers_mixed = receivers_mixed / np.max(np.abs(receivers_mixed))

# --- Export ---
output_path = wav_dir
sf.write(os.path.join(output_path, 'first_clean.wav'), in1, fs)
sf.write(os.path.join(output_path, 'second_clean.wav'), in2, fs)
sf.write(os.path.join(output_path, 'first_reverb.wav'), receiver_first, fs)
sf.write(os.path.join(output_path, 'second_reverb.wav'), receiver_second, fs)
sf.write(os.path.join(output_path, 'noise.wav'), noise_only, fs)
sf.write(os.path.join(output_path, 'mixed_output.wav'), receivers_mixed, fs)

print("Processing complete. Files saved using hParams configuration.")